src/kernel/task.zig

```diff
@@ -99,6 +104,37 @@ pub const Task = struct {
         return task;
     }
 
+    pub fn createFromElf(program_elf: elf.Elf, kernel: bool, task_vmm: *vmm.VirtualMemoryManager(arch.VmmPayload), allocator: *Allocator) (bitmap.Bitmap(usize).BitmapError || vmm.VmmError || elf.Error || Allocator.Error)!*Task {
+        const task = try create(program_elf.header.entry_address, kernel, task_vmm, allocator);
+        errdefer task.destroy(allocator);
+
+        // Iterate over sections
+        for (program_elf.section_headers) |header, i| {
+            if ((header.flags & elf.SECTION_ALLOCATABLE) == 0) {
+                continue;
+            }
+            errdefer {
+                // Free the previously allocated addresses
+                for (program_elf.section_headers) |header2, j| {
+                    if (j >= i)
+                        break;
+                    if ((header2.flags & elf.SECTION_ALLOCATABLE) != 0)
+                        task_vmm.free(header.virtual_address) catch |e| panic(null, "VMM failed to clean up a previously-allocated address after an error: {}\n", .{e});
+                }
+            }
+            // If it is loadable then allocate it at its virtual address
+            const attrs = vmm.Attributes{ .kernel = kernel, .writable = (header.flags & elf.SECTION_WRITABLE) != 0, .cachable = true };
+            const vmm_blocks = std.mem.alignForward(header.size, vmm.BLOCK_SIZE) / vmm.BLOCK_SIZE;
+            const vaddr_opt = try task_vmm.alloc(vmm_blocks, header.virtual_address, attrs);
+            const vaddr = vaddr_opt orelse return if (try task_vmm.isSet(header.virtual_address)) error.AlreadyAllocated else error.OutOfBounds;
+            if (vaddr != header.virtual_address) panic(null, "VMM didn't allocate at correct address. Expected 0x{X} but got 0x{X}\n", .{ header.virtual_address, vaddr });
```

Comment:
Is this possible as the alloc would fail/return null if it can't alloc at that address

Reply:
True :+1: